ssv/committee_runner.go

```diff
@@ -0,0 +1,482 @@
+	}
+
+	if !quorum {
+		return nil
+	}
+
+	// Get unique roots to avoid repetition
+	rootSet := make(map[[32]byte]struct{})
+	for _, root := range rootsList {
+		rootSet[root] = struct{}{}
+	}
+
+	// Get validator-root maps for attestations and sync committees, and the root-beacon object map
+	attestationMap, committeeMap, beaconObjects, err := cr.expectedPostConsensusRootsAndBeaconObjects()
+	if err != nil {
+		return errors.Wrap(err, "could not get expected post consensus roots and beacon objects")
+	}
+
+	var anyErr error
+	attestationsToSubmit := make(map[phase0.ValidatorIndex]*phase0.Attestation)
+	syncCommitteeMessagesToSubmit := make(map[phase0.ValidatorIndex]*altair.SyncCommitteeMessage)
+
+	// For each root that got at least one quorum, find the duties associated to it and try to submit
+	for root := range rootSet {
+
+		// Get validators related to the given root
+		role, validators, found := findValidators(root, attestationMap, committeeMap)
+
+		if !found {
+			// Check if duty has terminated (runner has submitted for all duties)
+			if cr.HasSubmittedAllValidatorDuties(attestationMap, committeeMap) {
+				cr.BaseRunner.State.Finished = true
+			}
+			// All roots have quorum, so if we can't find validators for a root, it means we have a bug
+			// We assume it is safe to stop due to honest majority assumption
+			return errors.New("could not find validators for root")
+		}
+
+		for _, validator := range validators {
+
+			// Skip if no quorum - We know that a root has quorum but not necessarily for the validator
+			if !cr.BaseRunner.State.PostConsensusContainer.HasQuorum(validator, root) {
+				continue
+			}
+			// Skip if already submitted
+			if cr.HasSubmitted(role, validator) {
+				continue
+			}
+
+			// Reconstruct signature
+			share := cr.BaseRunner.Share[validator]
+			pubKey := share.ValidatorPubKey
+			sig, err := cr.BaseRunner.State.ReconstructBeaconSig(cr.BaseRunner.State.PostConsensusContainer, root,
+				pubKey[:], validator)
+			// If the reconstructed signature verification failed, fall back to verifying each partial signature
+			if err != nil {
+				for root := range rootSet {
+					cr.BaseRunner.FallBackAndVerifyEachSignature(cr.BaseRunner.State.PostConsensusContainer, root,
+						share.Committee, validator)
+				}
+				// Record the error and continue to next validators
+				anyErr = errors.Wrap(err, "got post-consensus quorum but it has invalid signatures")
+				continue
+			}
+			specSig := phase0.BLSSignature{}
+			copy(specSig[:], sig)
+
+			// Get the beacon object related to root
+			if _, exists := beaconObjects[validator]; !exists {
+				anyErr = errors.Wrap(err, "could not find beacon object for validator")
+				continue
+			}
+			if _, exists := beaconObjects[validator][root]; !exists {
+				anyErr = errors.Wrap(err, "could not find beacon object for validator")
+				continue
+			}
+			sszObject := beaconObjects[validator][root]
+
```

Comment:
imo this looks better buts its a preference 
```suggestion
			validatorObjs, exists := beaconObjects[validator]
            if !exists {
				anyErr = errors.Wrap(err, "could not find beacon object for validator")
				continue
			}
			sszObject, rexists := validatorObjs[root]
            if !rexists {
				anyErr = errors.Wrap(err, "could not find beacon object for validator")
				continue
			}

```